File: backend/src/main/java/com/alumnihub/service/JobService.java
package com.alumnihub.service;

import com.alumnihub.dto.JobOpeningCreateDto;
import com.alumnihub.dto.JobOpeningDto;
import com.alumnihub.entity.JobOpening;
import com.alumnihub.entity.User;
import com.alumnihub.repository.JobOpeningRepository;
import com.alumnihub.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class JobService {

    private final JobOpeningRepository jobOpeningRepository;
    private final UserRepository userRepository;

    @Transactional
    public JobOpeningDto createJob(String userEmail, JobOpeningCreateDto createDto) {
        User creator = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        JobOpening job = JobOpening.builder()
                .company(createDto.getCompany())
                .role(createDto.getRole())
                .location(createDto.getLocation())
                .category(createDto.getCategory())
                .description(createDto.getDescription())
                .requirements(createDto.getRequirements())
                .externalLink(createDto.getExternalLink())
                .creator(creator)
                .build();

        JobOpening saved = jobOpeningRepository.save(job);
        return convertToDto(saved, creator);
    }

    public List<JobOpeningDto> getJobs(String userEmail) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        return jobOpeningRepository.findAll().stream()
                .map(j -> convertToDto(j, user))
                .collect(Collectors.toList());
    }

    @Transactional
    public void deleteJob(String userEmail, UUID jobId) {
        JobOpening job = jobOpeningRepository.findById(jobId)
                .orElseThrow(() -> new IllegalArgumentException("Job not found: " + jobId));

        if (!job.getCreator().getEmail().equalsIgnoreCase(userEmail)) {
            throw new org.springframework.security.access.AccessDeniedException("Not authorized to delete this job.");
        }

        jobOpeningRepository.delete(job);
    }

    @Transactional
    public JobOpeningDto saveJob(String userEmail, UUID jobId) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        JobOpening job = jobOpeningRepository.findById(jobId)
                .orElseThrow(() -> new IllegalArgumentException("Job not found: " + jobId));

        job.getSavedBy().add(user);
        JobOpening saved = jobOpeningRepository.save(job);
        return convertToDto(saved, user);
    }

    @Transactional
    public JobOpeningDto unsaveJob(String userEmail, UUID jobId) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        JobOpening job = jobOpeningRepository.findById(jobId)
                .orElseThrow(() -> new IllegalArgumentException("Job not found: " + jobId));

        job.getSavedBy().remove(user);
        JobOpening saved = jobOpeningRepository.save(job);
        return convertToDto(saved, user);
    }

    public List<JobOpeningDto> getSavedJobs(String userEmail) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        return jobOpeningRepository.findAll().stream()
                .filter(j -> j.getSavedBy().contains(user))
                .map(j -> convertToDto(j, user))
                .collect(Collectors.toList());
    }

    private JobOpeningDto convertToDto(JobOpening j, User user) {
        return JobOpeningDto.builder()
                .id(j.getId())
                .company(j.getCompany())
                .role(j.getRole())
                .location(j.getLocation())
                .category(j.getCategory())
                .description(j.getDescription())
                .requirements(j.getRequirements())
                .externalLink(j.getExternalLink())
                .creatorId(j.getCreator().getId())
                .creatorName(j.getCreator().getFullName())
                .saved(j.getSavedBy().contains(user))
                .build();
    }
}

File: frontend/services/jobService.ts
import { getAuthToken } from "./authService";

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080";

export interface JobOpeningDto {
  id: string;
  company: string;
  role: string;
  location?: string;
  category: string;
  description: string;
  requirements?: string;
  externalLink?: string;
  creatorId: string;
  creatorName: string;
  saved: boolean;
}

export interface JobOpeningCreateDto {
  company: string;
  role: string;
  location?: string;
  category: string;
  description: string;
  requirements?: string;
  externalLink?: string;
}

async function authFetch(url: string, options: RequestInit = {}) {
  const token = getAuthToken();
  if (!token) throw new Error("No authentication token found.");
  const res = await fetch(url, {
    ...options,
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json", ...options.headers },
  });
  if (!res.ok) throw new Error(await res.text() || "Request failed");
  return res;
}

export async function createJob(dto: JobOpeningCreateDto): Promise<JobOpeningDto> {
  const res = await authFetch(`${API_BASE}/api/jobs`, { method: "POST", body: JSON.stringify(dto) });
  return res.json();
}

export async function getJobs(): Promise<JobOpeningDto[]> {
  const res = await authFetch(`${API_BASE}/api/jobs`);
  return res.json();
}

export async function deleteJob(id: string): Promise<void> {
  await authFetch(`${API_BASE}/api/jobs/${id}`, { method: "DELETE" });
}

export async function saveJob(id: string): Promise<JobOpeningDto> {
  const res = await authFetch(`${API_BASE}/api/jobs/${id}/save`, { method: "POST" });
  return res.json();
}

export async function unsaveJob(id: string): Promise<JobOpeningDto> {
  const res = await authFetch(`${API_BASE}/api/jobs/${id}/unsave`, { method: "POST" });
  return res.json();
}

export async function getSavedJobs(): Promise<JobOpeningDto[]> {
  const res = await authFetch(`${API_BASE}/api/jobs/saved`);
  return res.json();
}
